ocaml/xapi/xapi_pool_helpers.ml

```diff
@@ -269,3 +269,221 @@ let apply_guest_agent_config ~__context =
+       day_configed=%Ld"
+      (frequency_to_str ~frequency)
+      day_configed ;
+    let day_configed_int = Int64.to_int day_configed in
+    let utc_now = Ptime_clock.now () in
+    let tz_offset_s = Ptime_clock.current_tz_offset_s () |> Option.get in
+    let seconds_in_a_day = seconds_random_within_a_day () in
+    let utc_start_of_next_sched_day =
+      utc_start_of_next_scheduled_day ~utc_now ~tz_offset_s ~frequency
+        ~day_configed_int
+    in
+    let delay =
+      update_sync_delay_for_next_schedule_internal ~utc_now
+        ~utc_start_of_next_sched_day ~seconds_in_a_day
+    in
+    print_next_schedule ~delay ~utc_now ~tz_offset_s ;
+    delay
+
+  let update_sync_delay_for_retry ~num_of_retries_for_last_scheduled_update_sync
+      =
+    let delay =
+      match num_of_retries_for_last_scheduled_update_sync with
+      | 1 ->
+          (* a random time between 1 hour and 2 hours *)
+          let secs_of_an_hour = 60. *. 60. in
+          secs_of_an_hour +. Random.float secs_of_an_hour
+      | 2 ->
+          (* a random time between 1 hour and 3 hours *)
+          let secs_of_an_hour = 60. *. 60. in
+          secs_of_an_hour +. (Random.float 2. *. secs_of_an_hour)
+      | 3 ->
+          (* a random time between 1 hour and 5 hours *)
+          let secs_of_an_hour = 60. *. 60. in
+          secs_of_an_hour +. (Random.float 4. *. secs_of_an_hour)
+      | n ->
+          raise (UpdateSync_RetryNumExceeded n)
+    in
+    let utc_now = Ptime_clock.now () in
+    let tz_offset_s = Ptime_clock.current_tz_offset_s () |> Option.get in
+    print_next_schedule ~delay ~utc_now ~tz_offset_s ;
+    delay
+
+  let rec update_sync ~num_of_retries_for_last_scheduled_update_sync =
+    Server_helpers.exec_with_new_task "periodic_update_sync" (fun __context ->
+        Helpers.call_api_functions ~__context (fun rpc session_id ->
+            if num_of_retries_for_last_scheduled_update_sync > 0 then
+              debug "[PeriodicUpdateSync] number of retries: %d"
+                num_of_retries_for_last_scheduled_update_sync ;
+            try
+              ignore
+                (Client.Pool.sync_updates ~rpc ~session_id
+                   ~self:(Helpers.get_pool ~__context)
+                   ~force:false ~token:"" ~token_id:""
+                ) ;
+              Xapi_periodic_scheduler.add_to_queue
+                periodic_update_sync_task_name Xapi_periodic_scheduler.OneShot
+                (update_sync_delay_for_next_schedule ~__context) (fun () ->
+                  update_sync ~num_of_retries_for_last_scheduled_update_sync:0
+              )
+            with _ ->
+              (* retry at most 3 times for each scheduled update sync *)
+              if num_of_retries_for_last_scheduled_update_sync < 3 then
+                try
+                  let num_str =
+                    match num_of_retries_for_last_scheduled_update_sync + 1 with
```

Comment:
Would suggest to do the retry in `Repository.sync`.
When it was design, it was assumed that the clients would perform the retry. But now, the call will be called by XAPI  itself, so it would make sense to get the call itself handle the retry.

Reply:
As we discussed offline, will remove the retry for failure, as it is a periodic update, any failure will be covered by next run. While since the update sync failure may out of user's expectation for the periodic behavior, will raise an alert when update sync failed.